src/nupic/algorithms/AnomalyLikelihood.hpp

```diff
@@ -0,0 +1,171 @@
+ *
+ * You should have received a copy of the GNU Affero Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses.
+ *
+ * http://numenta.org/licenses/
+ * ---------------------------------------------------------------------
+ */
+#ifndef NUPIC_ALGORITHMS_ANOMALY_LIKELIHOOD_HPP_
+#define NUPIC_ALGORITHMS_ANOMALY_LIKELIHOOD_HPP_
+
+#include <nupic/types/Types.hpp>
+#include <nupic/utils/MovingAverage.hpp>
+
+#include <string>
+#include <boost/circular_buffer.hpp>
+#include <vector>
+#include <cmath>
+#include <cassert>
+
+/**
+Note: this is an implementation from python in nupic repository.
+
+This module analyzes and estimates the distribution of averaged anomaly scores
+from a given model. Given a new anomaly score `s`, estimates `P(score >= s)`.
+
+The number `P(score >= s)` represents the likelihood of the current state of
+predictability. For example, a likelihood of 0.01 or 1% means we see this much
+predictability about one out of every 100 records. The number is not as unusual
+as it seems. For records that arrive every minute, this means once every hour
+and 40 minutes. A likelihood of 0.0001 or 0.01% means we see it once out of
+10,000 records, or about once every 7 days.
+
+USAGE
+-----
+
+There are two ways to use the code: using the AnomalyLikelihood helper class or
+using the raw individual functions.
+
+
+Helper Class
+------------
+The helper class AnomalyLikelihood is the easiest to use.  To use it simply
+create an instance and then feed it successive anomaly scores:
+
+anomalyLikelihood = AnomalyLikelihood()
+while still_have_data:
+  # Get anomaly score from model
+
+  # Compute probability that an anomaly has ocurred
+  anomalyProbability = anomalyLikelihood.anomalyProbability(
+      value, anomalyScore, timestamp)
+
+**/
+
+using namespace std;
+
+namespace nupic {
+  namespace algorithms {
+    namespace anomaly {
+
+struct DistributionParams {
+  DistributionParams(std::string name, Real32 mean, Real32 variance, Real32 stdev) :
+    name(name),mean(mean), variance(variance), stdev(stdev) {}
+  std::string name;
+  Real32 mean;
+  Real32 variance;
+  Real32 stdev;
+};
+
+class AnomalyLikelihood {
+
+  public:
+    AnomalyLikelihood(UInt learningPeriod=288, UInt estimationSamples=100, UInt historicWindowSize=8640, UInt reestimationPeriod=100, UInt aggregationWindow=10);
+
+    Real anomalyProbability(Real rawValue, Real anomalyScore, int timestamp=-1);
+
+  private:
+    //methods:
+    vector<Real> estimateAnomalyLikelihoods(vector<Real> anomalyScores, UInt skipRecords=0, UInt verbosity=0);
+    vector<Real>  updateAnomalyLikelihoods(vector<Real> anomalyScores, UInt verbosity=0);
+    Real32 tailProbability(Real32 x) const;
+    DistributionParams estimateNormal(vector<Real> sampleData, bool performLowerBoundCheck=true);
+    DistributionParams nullDistribution() const;
+
+    //private static methods
+static Real  computeLogLikelihood(Real likelihood)  {
```

Comment:
Just a quick note, why should this be public? The python file has only 2 high-level fuctions: constructor and anomalyProbabilities() (=compute()); the rest is described as "helper methods", so I thought I'd hide that implementation into `private:` (?)